pylint/config.py

```diff
@@ -79,39 +75,108 @@ def save_results(results, base):
         print('Unable to create file %s: %s' % (data_file, ex), file=sys.stderr)
 
 
-def find_pylintrc():
-    """search the pylint rc file and return its path if it find it, else None
+# TODO: Put into utils
+def walk_up(from_dir):
+    """Walk up a directory tree
+    :param from_dir: The directory to walk up from.
+        This directory is included in the output.
+    :type from_dir: str
+    :returns: Each parent directory
+    :rtype: generator(str)
+    """
+    cur_dir = None
+    new_dir = os.path.expanduser(from_dir)
+    new_dir = os.path.abspath(new_dir)
+
+    # The parent of the root directory is the root directory.
+    # Once we have reached it, we are done.
+    while cur_dir != new_dir:
+        cur_dir = new_dir
+        yield cur_dir
+        new_dir = os.path.abspath(os.path.join(cur_dir, os.pardir))
+
+
+def find_pylintrc_in(search_dir):
+    """Find a pylintrc file in the given directory.
+    :param search_dir: The directory to search.
+    :type search_dir: str
+    :returns: The path to the pylintrc file, if found.
+        Otherwise None.
+    :rtype: str or None
+    """
+    path = None
+
+    search_dir = os.path.expanduser(search_dir)
+    if os.path.isfile(os.path.join(search_dir, 'pylintrc')):
+        path = os.path.join(search_dir, 'pylintrc')
+    elif os.path.isfile(os.path.join(search_dir, '.pylintrc')):
+        path = os.path.join(search_dir, '.pylintrc')
+
+    return path
+
+
+def find_nearby_pylintrc(search_dir=''):
+    """Search for the nearest pylint rc file.
+    :param search_dir: The directory to search.
+    :type search_dir: str
+    :returns: The absolute path to the pylintrc file, if found.
+        Otherwise None
+    :rtype: str or None
     """
-    # is there a pylint rc file in the current directory ?
-    if os.path.exists('pylintrc'):
-        return os.path.abspath('pylintrc')
-    if os.path.exists('.pylintrc'):
-        return os.path.abspath('.pylintrc')
-    if os.path.isfile('__init__.py'):
-        curdir = os.path.abspath(os.getcwd())
-        while os.path.isfile(os.path.join(curdir, '__init__.py')):
-            curdir = os.path.abspath(os.path.join(curdir, '..'))
-            if os.path.isfile(os.path.join(curdir, 'pylintrc')):
-                return os.path.join(curdir, 'pylintrc')
-            if os.path.isfile(os.path.join(curdir, '.pylintrc')):
-                return os.path.join(curdir, '.pylintrc')
-    if 'PYLINTRC' in os.environ and os.path.exists(os.environ['PYLINTRC']):
+    search_dir = os.path.expanduser(search_dir)
+    path = find_pylintrc_in(search_dir)
+
+    for search_dir in walk_up(search_dir):
+        if path or not os.path.isfile(os.path.join(search_dir, '__init__.py')):
```

Comment:
How will this work for Python 3.3+ namespace packages? From my understanding, it seems this change won't support them. I'm wondering if we should though, what is your opinion on it?

Reply:
You're right, it won't. We could do something like; once we don't find an `__init__.py`, keep looking for the closest pylintrc file and only use it if a find a setup.py in a parent directory. I'm not sure how I feel about looking for `setup.py` files because it's possible that the package isn't installed that way, but I can't think of another way of doing it.

Reply:
Yeah, I feel the same way about a setup.py. Maybe expose a limit of levels, e.g. lookup maximum 5 levels, if there is no pylintrc in any of those, bail out with an exception?

Reply:
We'll have a global/commond line config by this point so that should work!